bookwyrm/models/bookwyrm_import_job.py

```diff
@@ -85,274 +83,147 @@ def process_books(job, tar):
-        "authors",
-        "readthroughs",
-        "shelves",
-        "shelf_books",
-        "lists",
-        "list_items",
-        "reviews",
-        "comments",
-        "quotes",
-    ]
-
-    for val in associated_values:
-        del book[val]
-
-    # now we can save the book as an Edition
-    new_book = models.Edition.objects.create(**book)
-    new_book.authors.set(local_authors)  # now we can add authors with set()
-
-    # get cover from original book_data because we lost it in clean_values
-    if cover_path:
-        tar.write_image_to_file(cover_path, new_book.cover)
-
-    # NOTE: clean_values removes "last_edited_by"
-    # because it's a user ID from the old database
-    # if this is required, bookwyrm_export_job will
-    # need to bring in the user who edited it.
-
-    # create parent
-    work = models.Work.objects.create(title=book["title"])
-    work.authors.set(local_authors)
-    new_book.parent_work = work
-
-    new_book.save(broadcast=False)
-    return new_book
-
-
-def clean_values(data):
-    """clean values we don't want when creating new instances"""
-
-    values = [
-        "id",
-        "pk",
-        "remote_id",
-        "cover",
-        "preview_image",
-        "last_edited_by",
-        "last_edited_by_id",
-        "user",
-        "book_list",
-        "shelf_book",
-        "parent_work_id",
-    ]
-
-    common = data.keys() & values
-    new_data = data
-    for val in common:
-        del new_data[val]
-    return new_data
-
-
-def find_existing(cls, data):
-    """Given a book or author, find any existing model instances"""
-
-    identifiers = [
-        "openlibrary_key",
-        "inventaire_id",
-        "librarything_key",
-        "goodreads_key",
-        "asin",
-        "isfdb",
-        "isbn_10",
-        "isbn_13",
-        "oclc_number",
-        "origin_id",
-        "viaf",
-        "wikipedia_link",
-        "isni",
-        "gutenberg_id",
-    ]
-
-    match_fields = []
-    for i in identifiers:
-        if data.get(i) not in [None, ""]:
-            match_fields.append({i: data.get(i)})
-
-    if len(match_fields) > 0:
-        match = cls.objects.filter(reduce(operator.or_, (Q(**f) for f in match_fields)))
-        return match.first()
-    return None
-
-
-def get_or_create_authors(data):
-    """Take a JSON string of authors find or create the authors
-    in the database and return a list of author instances"""
-
+    # make sure we have the authors in the local DB
     authors = []
-    for author in data:
-        clean = clean_values(author)
-        existing = find_existing(models.Author, clean)
+    for author in book_data.get("authors"):
+        existing = models.Author.find_existing(author)
         if existing:
             authors.append(existing)
         else:
-            new = models.Author.objects.create(**clean)
-            authors.append(new)
-    return authors
+            ap_author = activitypub.base_activity.ActivityObject(**author)
```

Comment:
Does `to_model()` work if switching to the concrete class the data refers to? As in:

```suggestion
            ap_author = activitypub.book.Author(**author)
```

This way, that `ap_author` object is built from a definition that knows what all the fields are, and will populate them with the defaults.

Reply:
This is the test I used:

```py
    def test_aut(self, *_):
        author = models.Author.objects.create(name="Author Name")
        apjson = author.to_activity()
        ap_author = activitypub.book.Author(**apjson)
        instance = ap_author.to_model(model=models.Author, save=True)
        self.assertEqual(instance, author)
```

Reply:
hmm, that does kind of make more sense but doesn't appear to solve the problem.

I've looked into this a bit more and it seems we should use `activitypub.parse(author)` etc. Might need to pull in parent work data for the book but that's kind of a separate issue. Thanks for prompting me to think about this differently @dato